### ISSUE TITLE
Roadmap and ideas for Molly.jl development

### ISSUE THREAD

**Issue by jgreener64** [jgreener64/Molly.jl#2]
This issue is a roadmap for Molly.jl development. Feel free to discuss things here or submit a PR. Bear in mind that significant refactoring will probably occur as the package develops.

### Low hanging fruit

Want to get involved? These issues might be the place to start:
- Speedups to the code. Both on the level of individual functions and the overall algorithm.
- Look over the code to find mistakes in implementation. There will be some because...
- Test coverage is minimal. We need tests for the 'core' functions that are unlikely to change their return values, e.g. the force functions.

### Bigger things to tackle

- Energy minimisation by gradient descent.
- Add other forcefields.
- Add other thermostats/ensembles.
- Get it working on the GPU.
- Line by line checking of how a mature MD package works to see where we have missed things and to get an idea of how to organise the codebase in future.
- Neighbour list in a cubic lattice, and use of group lists.
- Setup of MD simulations, e.g. adding hydrogen, solvent.
- Trajectory/topology file format readers and writers. The current topology reader is not very robust.

### Projects that will require more planning or will cause significant changes

- Abstracting the forcefield from the integration to allow different forcefields and integrators to be combined.
- Allow definition of custom forcefields.
- Parallelisation.

Okay, so this is less of a roadmap and more of a list of ideas to work on. But most or all of the above would be required to make this a package suitable for general use, so in a sense this is the roadmap.

**Comment by jgreener64** [comment]
"Abstracting the forcefield from the integration to allow different forcefields and integrators to be combined" and "Allow definition of custom forcefields" are now done.

**Comment by jarvist** [comment]
It may be interesting to see whether you can either directly use https://github.com/libAtoms/JuLIP.jl as a library (also https://github.com/libAtoms/NeighbourLists.jl), or at least be inspired by some of the code therein.

**Comment by jgreener64** [comment]
Thanks Jarvist, I have had a brief look at the code there but will dive a bit deeper.

**Comment by Zarasthustra** [comment]
Awhile ago I wrote some Julia code for EWALDS. It is verified for energy calculations against a NIST database.

https://www.nist.gov/mml/csd/chemical-informatics-research-group/spce-water-reference-calculations-10%C3%A5-cutoff
[https://www.nist.gov/sites/default/files/styles/480_x_480_limit/public/images/mml/csd/informatics_research/spce_schematic_3.jpg?itok=OTb11Rb_]<https://www.nist.gov/mml/csd/chemical-informatics-research-group/spce-water-reference-calculations-10%C3%A5-cutoff>
SPC/E Water Reference Calculations - 10Å cutoff | NIST<https://www.nist.gov/mml/csd/chemical-informatics-research-group/spce-water-reference-calculations-10%C3%A5-cutoff>
In this section, we provide sample configurations of SPC/E Water molecules[1] and report the various energy and force calculations for those configurations, where the coulombic contributions are computed using the traditional Ewald Summation Method[2]. These sample configurations and reference ...
www.nist.gov


Would you be interested in it? I have not taken its derivative yet to calculate force. It has been on my list of things to do, but I just don't seem to have time :S

I have also added Rattle to my code for bond constraints... it is fairly straight forward too. Once your code has EWALD and Rattle, it will be a "full" code in my opinion. You would probably want to upgrade it to particle mesh ewald for MD, but that is purely performance.

Braden
________________________________

**Comment by jgreener64** [comment]
> Would you be interested in it?
Yes, definitely. If you could open a PR when you get a chance that would be great. I understand how it's hard to find time though, I only seem to get a few hours a month on this package these days.

**Comment by jgreener64** [comment]
There is a Julia interface to [Chemfiles](https://github.com/chemfiles/chemfiles) - building Molly on top of those types would allow easy reading and writing.

**Comment by EricAlcaide** [comment]
what are your thoughts on GPU acceleration or at least multi-core in CPU? In python it's really easy with joblib, cupy, jax... Don't know about the julia set of tools for that but might be worth an exploration

**Comment by jgreener64** [comment]
GPU acceleration is feasible using [CuArrays.jl](https://github.com/JuliaGPU/CuArrays.jl) - it would probably involve refactoring into a style that uses broadcasting rather than for loops, but that's no bad thing anyway.

Multi-core CPU support has improved greatly in the recent Julia v1.3 release so this is another avenue to explore, though I personally don't have much experience there.

Another point to add here is the ability to differentiate through molecular simulations using [Zygote.jl](https://github.com/FluxML/Zygote.jl), which is an exciting use case and something that Julia is well-suited for.

**Comment by cortner** [comment]
I was unfortunately unaware of this package until now. (Or maybe I saw it and have forgotten...) Anyhow, I  wonder whether there is any chance to combine Molly and JuLIP, or at least make them compatible.

**Comment by cortner** [comment]
Regarding the neighbourlist - this is surely something that we can share. I believe my list is fairly fast, it is based on `matscipy`. The main downside for Molly is that it uses the same cutoff for all atoms. It would need to be generalised so that different atoms can have a different cutoff. 

I‘ve also started playing around with implementing various iterators over the neighbourlist to have a more functional style in implementing potentials. I‘m less sure how well that worked.

**Comment by jgreener64** [comment]
Thanks for commenting @cortner. Once I get some time for this package, diving into JuLIP is definitely on the list of priorities to see what can be combined. Certainly with things you have optimised, like the neighbour list, there is room for Molly to use or build on that.

**Comment by cortner** [comment]
or extract shared functionality so that JuLIP and Molly become effectively frontends...

**Comment by AlfTetzlaff** [comment]
Hi! I observed some things in your code which could be optimized or written the "julia way".
1) Use [StaticArrays](https://github.com/JuliaArrays/StaticArrays.jl) to store positions and velocities, instead of your self-defined mutable structs. This might speed things up quite a lot due to allowing SIMD.
2) Implement a dist(a,b,box) function specialized for static arrays which takes the boundary conditions into account.
3) r2 = dx*dx+... can then be written as r^2. The pow function has specializations for 1,2 and 3, so this produces no overhead compared to C.
4) If you're using Atom/Juno you can use the Juno.@profiler to find bottlenecks
5) There is a discourse thread about "[struct of arrays vs array of structs](https://discourse.julialang.org/t/struct-of-arrays-soa-vs-array-of-structs-aos/30015/18)". I dont remember who won, but maybe it's instructive to read.
6) @inline the update_accelerations! functions
7) Use structs with parametric types, so that you can easily switch to e.g Float32 for GPUs. (SVectors are parametric.)
8) In theory, just an idea, one could leave the time-integration to another package like DifferentialEquations. They provide a whole lot of explicit or implicit solvers (but velocity verlet is missing I think) with adaptable or fixed time steps.

Kind regards, Max

**Comment by cortner** [comment]
Some of this I do in JuLIP, though not yet as well as I’d like, it needs more cleanup to get proper type-agnostic code eg.

**Comment by jgreener64** [comment]
Thanks for the feedback @AlfTetzlaff, I would definitely like to get some of this in. I have tried a few of the suggestions previously, for example the positions and velocities are currently mutable static vectors (sub-types of `FieldVector`) which should hopefully be giving some of the speedups you talk about.

The types should be made parametric and it would help with GPU compatibility, though at the minute the code is not written in a GPU-friendly way so that would need to change first. I would like to use the DifferentialEquations integrators in the long-term since they make so many available, but for the proof-of-concept here I just implemented a simple velocity Verlet.

I have opened a [Discourse post](https://discourse.julialang.org/t/help-make-my-ideal-gas-simulation-ideally-fast/39141) with a specific MWE based on the current non-bonded force code to solicit feedback - please feel free to comment more there.

**Comment by Zarasthustra** [comment]
I did a whole bunch of timing on many different possible LJ calculations for force and at the end of it, using SVector was the fastest.

Also, simple things like image separation could be done in a vectorized for loop rather than sequentially, and there was some time saving there too. Not enough to restructure your code just for that, but none the less, some.

I have put my Molecular Dynamics coding on hold and have started doing Monte Carlo (I just like MC better, and find statistical mechanics more interesting).

So far my MC code can do rigid SPC/E water. So, somewhat of a toy system as well, but, It can do Wolf summation or Ewalds for the electrostatic. There is nothing toy about Ewalds. Once I have it cleaned up, I will upload it and share with you guys. The most important thing is that it uses OffsetArrays.jl which is a wonderful package. It allows indices to take on any value, not just 1 and up. This just makes the fourier piece nice since the lattice vectors range from -kmax:kmax.

Julia does Ewalds really well. The fourier part is bloody fast and the "hard" part if there is one. The real part is simple and similar to a bare coulomb calculation.

I am also planning on making some ewald unit tests for energy. Force would also be nice, but that is in the future.

So, now that I am back part time on the Julia wagon, I will try the ewald contributions to you guys soon.

NOTE: --------------------------------------
Here is some fortran code that does force ewald force routines. It will take a good day to go over what they did and figure out how to make it work, which is why I have been putting off the force calculation and stuck with the energy.

http://zeolites.cqe.northwestern.edu/Music/electrostatic/ewald.htm
zeolites.cqe.northwestern.edu<http://zeolites.cqe.northwestern.edu/Music/electrostatic/ewald.htm>
This modules contains routines for those nasty ewald summations.! Good reference for ewald: Allen and Tildesley, Computer Simulation of ! Liquids, p 155-162; Amit's notes for doing part of the calculation in the
zeolites.cqe.northwestern.edu


Braden
________________________________

**Comment by jgreener64** [comment]
Great, thanks for the input.

**Comment by jgreener64** [comment]
Here's a sneak preview of some non-bonded force optimisations I've been working on: faster single-CPU code, a multithreaded CPU implementation and a GPU implementation.

Hoping to refactor the package to allow these from the same API within a couple of weeks.

![image](https://user-images.githubusercontent.com/6620652/81618037-7305bc80-93de-11ea-9be3-5f1a6cb85746.png)

**Comment by AlfTetzlaff** [comment]
Could it be that you confused us with ms in the above graph by accident?

**Comment by jgreener64** [comment]
It's ms but for a 1000-step simulation, so it would be us per step. The blue dot for 100 atoms is the simulation benchmark time from the Discourse post.

**Comment by jgreener64** [comment]
Some changes I added this weekend:
- `SVector`s for coordinates and velocities, allowing arbitrary simulation dimensions.
- Multithreading of non-bonded force calculations.

**Comment by AlfTetzlaff** [comment]
Cool! Have you noticed any single threaded speedup due to SVectors?

**Comment by jgreener64** [comment]
Yes, the Discourse thread benchmark went from 45 ms to 36 ms using `SVector`s, rather than the mutable `FieldVector`s used before.

Making `Simulation` immutable also sped things up.

The cell neighbour list would be a big algorithmic speedup (as discussed above).

**Comment by jgreener64** [comment]
My rough feature list for a v0.1 release is:

- Cell neighbour list.
- Type parameterising (to allow Float32).
- A few more thermostats.
- A few more simulators, e.g. minimisation/Langevin dynamics.
- Some unit tests.
- More thorough docs and docstrings.

**Comment by Chenghao-Wu** [comment]
May I have your thoughts on the parallelization via domain decomposition in Julia?

**Comment by jgreener64** [comment]
It's not really my area of expertise, but I guess it can be done and it looks like it could be a useful algorithm. I don't know how well that type of simulation would work in the Molly framework currently though.

**Comment by cortner** [comment]
It would be an interesting proposition and a small step towards replacing lammps

**Comment by jgreener64** [comment]
I thought I'd just update this issue with the state of the project.

In the last few months Molly has been under heavy development and has seen many new features including Unitful.jl support, more interactions, an OpenMM XML force field reader, [AtomsBase.jl](https://github.com/JuliaMolSim/AtomsBase.jl) integration, differentiable simulation with Zygote.jl on CPU and GPU, a new general interaction type, implicit solvation models, more simulators, energy minimisation, more analysis functions and better test coverage. See the [release notes](https://github.com/JuliaMolSim/Molly.jl/blob/master/NEWS.md) for more details.

Future development directions include:
- Optimisation of the GPU code path to make performance competitive to related software. https://github.com/JuliaMolSim/Molly.jl/issues/60 discusses one way to do this, but any approach will probably use GPU kernels. The challenge is to make the kernels work with gradients and to maintain flexibility while allowing the specialisations required for high performance.
- More simulators such as replica exchange MD and Langevin variants.
- Support for constrained bond lengths and angles, which are used in most macromolecular simulations these days.
- More options and testing for simulation setup including residue template matching.
- Pressure coupling methods for NPT simulation.
- Support for non-cubic boundary conditions and no periodic boundaries in given dimensions.
- Particle mesh Ewald for long-range summation of electrostatic interactions.
- Parallel domain decomposition, or at least better use of CPU parallelism.

A few groups are interested in using Molly for a diverse range of simulations, so the intention is that it will stay general enough to be widely useful whilst also reaching high performance. My personal research uses Molly for differentiable simulations of proteins, so this will continue to dictate the features I work on.

**Comment by uliano** [comment]
May be of interest  a native Julia (as in no-external-dependencies) .xtc reader and writer? 
It is twice as fast than c implementation used by mdtraj.

**Comment by jgreener64** [comment]
That would be great. Currently we are using Chemfiles for that kind of thing but a fast native Julia approach would be welcome. It could even be its own package?

**Comment by uliano** [comment]
At the moment it is just a file with a bunch of functions. 

It started as a proof of concept, learning Julia, while showing its potential to both speed and low level pesky tasks with the *far future* eventual goal to assemble something resembling the bad copy of MDAnalysis or MDtraj.

I'm, at the moment, too ignorant of Julia packaging system, and of git and github machinery, to be able to provide it. However with some help I'd be glad to contribute. 

My original goal was to be able to natively read and write trajectory in the formats I use, namely gromacs .xtc, charmm namd .dcd, amber .nc and desmond .dtr. After the unexpected (at least speedwise) success with .xtc I got quickly stranded by netcdf3 underlying .nc and it all went on temporary hold.

The the functions are pretty solid as I tested low level ones against the c implementation while I was able to read and write real .xtc file with no diff from original and to compare the read coordinates with MDtraj's. It would be a pity not to use them, also given the effort done in understanding and optimizing the original c code (c implementation slowness is *obviously* due to implementation itself and not to the language)

I'm writing too much here is the thing

https://github.com/uliano/MDTools.jl

I'm willing to help in my spare time but keep in mind that at the moment I'm not proficient enough in both Julia and github packages machinery to be autonomous.

**Comment by jarvist** [comment]
I think it looks pretty good! It wouldn't be too much effort to rename your repo 'xtcreader.jl' or similar & then add the boilerplate to make it an independent package.

**Comment by uliano** [comment]
It actually writes, also!

**Comment by uliano** [comment]
So I did my homework and now we have a Julia package with xtc read and write functionality

https://github.com/uliano/MDTrajectoryFiles.jl

I added minimal docstrings on the exported functions and a very basic test reads and writes then compares. I will eventually implement other file formats so I choose a more generic name than XTC.

**Comment by uliano** [comment]
This is the last comment for a while, I promise!
I just wanted to say that MDTrajectoryFiles.jl has been accepted in the Julia registry. 

Foreseeable changes in the next few weeks:
- native Julia read & write of CHARMM/NAMD .dcd files
- possibility to load a subset of atoms from each frame.

**Comment by CoryKornowicz** [comment]
Using Metal.jl, I was able to swap all the CuArrays to MtlArrays for better performance on Apple M-series chips. Might be something to look into as a 'low hanging fruit.'

**Comment by jgreener64** [comment]
That's interesting, thanks. Once the `kernels` branch is merged (with >10x speedups on GPU) it would be worth allowing and documenting this.

See https://github.com/JuliaMolSim/Molly.jl/pull/99 for work towards generic array types.

**Comment by jgreener64** [comment]
Yes, see the `kernels` branch for a lot of progress on this. It uses Enzyme.jl for AD in the slow parts and writes GPU kernels, increasing performance and memory utilisation dramatically.

Hoping to merge and release soon, just waiting on some docs cleanup and a new Enzyme.jl release.

**Comment by jgreener64** [comment]
v0.15.0 is out with many improvements including GPU speed and memory usage.

Priorities for future development are:
- High GPU performance.
- Particle mesh Ewald summation.
- Constraints.
- Pressure coupling, this will require the concept of molecules to be added as molecules need to be moved together.
- More robust and flexible setup from files, essential for biomolecular systems.

Hopefully there will be a GSoC student(s) working on Molly this summer too.

**Comment by jgreener64** [comment]
Pressure coupling added via the Monte Carlo barostat.

@JaydevSR will be looking at GPU performance this summer as part of JSoC.

**Comment by jgreener64** [comment]
v0.22.0 is a big release with improved performance, especially on GPU, and support for other GPU backends.

See the release notes at https://github.com/JuliaMolSim/Molly.jl/releases/tag/v0.22.0.

**Comment by jgreener64** [comment]
Lots of work is ongoing for v1.0.0, hoping to get there in the next few months.

**Comment by jarvist** [comment]
@IngvarsV has done a lot of working integrating non-mean-field electrostatics (following Cahill, https://github.com/Frost-group/MembraneElectrostatics.jl , and also some work at Imperial by Alexei Kornyshev + group defining freq-dependent dielectric environments for more accurate implicit solvent models: Bessel functions everywhere, as far as the eye can see!) with Molly: 
https://github.com/Frost-group/BIOMD-Cahill/blob/master/Amino_Acid_Chains_2025/src/Interactions.jl

I'm not sure if there are any architectural choices in 'upstream' Molly that would make this easier / nicer, but perhaps it might be worth having a chat over the summer...

**Comment by jgreener64** [comment]
Great, yes it would be good to chat. Hopefully v1 should give you more stability to build on top of, but we want to get the design right.

**Comment by EuclidDivisionLemma** [comment]
Hello. Although my major is computer science, I've recently become interested in molecular theory. Do I need to have in-depth knowledge to understand the codebase? I thought I could learn it simultaneously while contributing if possible.

**Comment by jgreener64** [comment]
Give it a go and see, hopefully you should be able to get into the code.

**Comment by EuclidDivisionLemma** [comment]
Thanks. Do you have any suggestions/ideas for features so that I know exactly what to learn, as I'm new here? Most of the plans listed above are marked complete. I'm mostly interested in chemistry/physics-related stuff, as I can learn something new.

**Comment by jgreener64** [comment]
Have a browse of the code and docs and see what interests you, I might be able to suggest something based off that.

**Comment by EuclidDivisionLemma** [comment]
I was looking at the potential functions and found that most pair potentials have already been implemented. Sutherland's potential barely has any resources for me to learn. So I'm thinking of [the Axilrod–Teller potential](https://en.wikipedia.org/wiki/Axilrod%E2%80%93Teller_potential).
